proof/refine/ARM/Tcb_R.thy

```diff
@@ -3097,100 +3144,176 @@ lemma tc_corres_sched:
-                         cteDelete_invs' cteDelete_typ_at'_lifts cteDelete_sch_act_simple)+
-  apply (clarsimp simp: tcb_cte_cases_def cte_level_bits_def objBits_defs tcbIPCBufferSlot_def)
-  by (auto dest!: isCapDs isReplyCapD isValidVTableRootD simp: isCap_simps)
-  *)
+  apply (wpsimp wp: hoare_vcg_all_lift hoare_weak_lift_imp setP_invs'
+                    installTCBCap_invs' installThreadBuffer_invs' installTCBCap_sch_act_simple
+                    setMCPriority_invs' )
+  apply (clarsimp cong: conj_cong)
+  apply (fastforce simp: isValidFaultHandler_def isCap_simps isValidVTableRoot_def)
+  done
+
+lemma schedContextBindTCB_invs'_helper:
+  "\<lbrace>\<lambda>s. invs' s \<and>  ex_nonz_cap_to' tcbPtr s \<and> ex_nonz_cap_to' scPtr s \<and>
+        bound_sc_tcb_at' (\<lambda>sc. sc = None) tcbPtr s \<and> ko_at' sc scPtr s \<and> scTCB sc = None\<rbrace>
+   do y <- threadSet (tcbSchedContext_update (\<lambda>_. Some scPtr)) tcbPtr;
+      setSchedContext scPtr (scTCB_update (\<lambda>_. Some tcbPtr) sc)
+   od
+   \<lbrace>\<lambda>_. invs'\<rbrace>"
+  apply (simp add: schedContextUnbindTCB_def invs'_def valid_state'_def valid_pspace'_def)
+  apply (wp threadSet_valid_objs' threadSet_mdb' threadSet_sch_act threadSet_iflive'
+            threadSet_cap_to threadSet_ifunsafe'T  threadSet_idle'T threadSet_ctes_ofT
+            threadSet_valid_queues_new threadSet_valid_queues' threadSet_valid_release_queue
+            threadSet_valid_release_queue' threadSet_not_inQ threadSet_cur
+            untyped_ranges_zero_lift valid_irq_node_lift valid_irq_handlers_lift''
+            threadSet_ct_idle_or_in_cur_domain' hoare_vcg_const_imp_lift
+         | clarsimp simp: tcb_cte_cases_def cteCaps_of_def)+
+  apply (fastforce simp: valid_idle'_def idle_tcb'_2_def pred_tcb_at'_def obj_at'_def projectKOs
+                         objBits_def objBitsKO_def valid_tcb'_def tcb_cte_cases_def comp_def
+                         valid_obj'_def valid_sched_context'_def valid_sched_context_size'_def
+                         valid_release_queue'_def inQ_def
+                   elim: ps_clear_domE valid_objsE'[where obj="KOSchedContext sc"] split: if_splits)
+  done
+
+lemma schedContextBindTCB_invs':
+  "\<lbrace>\<lambda>s. invs' s \<and> ex_nonz_cap_to' tcbPtr s \<and> ex_nonz_cap_to' scPtr s \<and>
+        (ksSchedulerAction s = ResumeCurrentThread \<longrightarrow> ksCurThread s \<noteq> tcbPtr) \<and>
+        bound_sc_tcb_at' (\<lambda>sc. sc = None) tcbPtr s \<and> obj_at' (\<lambda>sc. scTCB sc = None) scPtr s\<rbrace>
+   schedContextBindTCB scPtr tcbPtr
+   \<lbrace>\<lambda>_. invs'\<rbrace>"
+  apply (simp add: schedContextBindTCB_def)
+  apply (subst bind_assoc[symmetric, where m="threadSet _ _"])
+  apply (rule hoare_seq_ext)+
+      apply wpsimp
+     apply (wpsimp wp: isSchedulable_wp)
+    apply (clarsimp simp: isSchedulable_bool_runnableE)
+    apply (wp (once) hoare_drop_imps, wp (once) hoare_drop_imps, wp hoare_vcg_imp_lift')
```

Comment:
That's a bit of an unpleasant pattern, but I guess acceptable if it's necessary, and it does look necessary. It would be nice to avoid this, but I don't think any of my usual tricks with `hoare_drop_imps` would apply.

Reply:
A possible alternative would be using `hoare_strengthen_post` to rewrite the postconditions to what you need.